### ISSUE TITLE
Cache full encode data

### ISSUE THREAD

**Issue by Frooxius** [Frooxius/NeosPublic#223]
Currently anytime somebody joins, the entire data model is serialized and transmitted over, which results in a tiny hiccup. This could be eliminated by caching the full encode data and invalidating individual elements on delta changes - the next encode would only have to re-encode elements that have been changed, significantly reducing the necessary workload.

However this isn't currently a big problem, as the hiccup is usually tiny enough (and could be further reduced with multi-threading in this part as well as change to threading model). However with more complex worlds (and thus data models) and significantly more users joining and leaving, this would help increase the throughput and pauses.

**Comment by TehTurk** [comment]
Appreciate you pointing me in this direction for the issue, but to answer your questions from the other thread.
- Doesn't matter, it seems to be more persistent with more users in a session, but headless does seem to handle it worse. 
- In context of Headless it's on a xeon v1234 with 16gb ram as for my main pc it's a Zen 3700x with 16gb as well. So nothing I would think to do with hardware. 
- Pretty sure it affects mainly host, but then propagate to users as they have to also wait for the 2 seconds on the host to adjust.
- Good way to test this would be to have a mass join event occur but as the player count increases, so does the time of halt/freeze.
- I can also get other users testimonies if need be as this has been someone of a commonly accepted thing among the community but I don't think it was made as a proper report until now, which I did in mindset of "Oh, someone should probrabbly write it down"

**Comment by Frooxius** [comment]
- That's really bizarre that headless seems to handle it worse. It has much better runtime and less overhead than the normal client. I've always had it behave much better in this regard. What makes you say that it handles it worse? I suspect the issue might actually lie elsewhere in your case if this is true
- You say that it propagates to other users. What exactly does that mean? What do those users experience? 

I've done some testing, where I put artificial 10 second delay before encoding full state whenever someone joins and had two clients connected. They still saw each other moving and voice was working perfectly, even if the host was frozen for 10 seconds. This is because the logic is handled on separate thread.

Any host-driven logic will however freeze, as the host stops doing updates.

**Comment by TehTurk** [comment]
- Hmm maybe it could be my user connection and host connection are on the same network? (User/Headless)
- Hmm I don't know because oft not if I'm hosting the session I'm locked up for 1-2 seconds before I can see what's happening, part of this is extrapolation and filling in the dots from what I've experienced it in other sessions on people joining the host. 
- It is only this severe/dramatic when the user count in a session is hitting higher limits like 10-20. 

Hmm okay, I'm aware of the thread being separate because I have had users waiting to join and the issue isn't super consistent, maybe I'm attributing the lag to not be related to network and could be something more. Maybe Avatar Load?

**Comment by Frooxius** [comment]
Yes it sounds like it's connection in your case. If it's the same thing that happened in your spooky world, that doesn't happen because the host locks up, that will only affect any logic run on host, but voice and positions should stay normal for all users.

It might be just your connection not handling the sudden burst of data when someone's connecting too well. Potentially it could be system resource starvation, but with the HW it seems unlikely to be the case.

This optimization will help prevent the host from locking up when someone connects, but it won't help with any networking part of it, that one is a bit more tricky to handle if the connection can handle the bandwidth.

**Comment by TehTurk** [comment]
Dang okay thanks for the feedback Froox, I'll try to monitor my network and see if I can get something more tangible data wise, because I know my data speeds bandwidth wise should be enough but maybe it's because it's doing extra work communication wise because it's a net connect, not a lan connect. Because I've definitely had it more consistently when I'm connected to my headless.

**Comment by Frooxius** [comment]
Yes it sounds like it's connection in your case. If it's the same thing that happened in your spooky world, that doesn't happen because the host locks up, that will only affect any logic run on host, but voice and positions should stay normal for all users.

It might be just your connection not handling the sudden burst of data when someone's connecting too well. Potentially it could be system resource starvation, but with the HW it seems unlikely to be the case.

This optimization will help prevent the host from locking up when someone connects, but it won't help with any networking part of it, that one is a bit more tricky to handle if the connection can handle the bandwidth.

**Comment by Electronus** [comment]
So this may be a misunderstanding on my part but even when im not host I freeze when someone joins a session, often I can tell who it is, by how long I freeze.

**Comment by TehTurk** [comment]
That's due to the host getting the effect and then propagating to the end users (you) in this case. Because the way networking works, everything that the host sees, the end user or connected user will see that.

**Comment by Electronus** [comment]
@TehTurk Are you sure? My understanding is the World would appear to freeze but peer-to-peer and rendering would continue, but in my case the whole program freezes not drawing a  new frame until the user is at least partially loaded.

**Comment by Frooxius** [comment]
It depends on the type of the freeze. If you see everyone freeze for a bit, but you can still move around and the rendering continues, that's host hard freezing, because that briefly stops them relaying all the traffic.

If your rendering freezes, that's because of a local process - e.g. heavy local processing, asset loading triggering full GC pause and things like that. That's unrelated to the host and this issue, as it will not help with that kind of freeze.

**Comment by Electronus** [comment]
Okay so separate issue then? right will have another look for a better one or make one when it isn't so hot. Never mind me then!

**Comment by Frooxius** [comment]
I wouldn't recommend issue for that unless it's some specific cause. Otherwise it's more of a "doing lots of heavy things lags", which is not very actionable on itself as it covers lots of different things and it's handled more as part of general profiling and performance optimizations.

We do have a number of performance optimizations coming that will help reduce various sources of this - LogiX, asset loading and so on.

**Comment by shiftyscales** [comment]
@Frooxius if I'm understanding the issue correctly, this was one of the largest problems experienced by most during VBLFC. With fairly complex data models, and larger user counts, the headless client would pause anytime a user joined, resulting in a few second break where data stopped being transmitted followed by a catch up.

In a social environment it was fairly disruptive. It isn't as necessary right now without an event on as the average user turnover of a session is less, but this is definitely something we should have implemented in time for any other larger-scale events hosted in Neos/it should be given priority if such an event opportunity comes up.
